### DISCUSSION TITLE
Is it possible to add vocabulary to a clip model packaged inside of a distributed SD checkpoint?

### DISCUSSION THREAD

**Discussion by TeKett** [huggingface/diffusers#6876] [Q&A]
Is it possible to add vocabulary to a clip model packaged inside of a distributed SD checkpoint? I have a checkpoint of the .safetensor variety that i downloaded on civitai, that i want to train/finetune and make loras for. But i want to add new vocabulary to it in order to improve training, and not have the tokens split into other words, and damage things it already knows. To make it easy for me i wanted to add the booru tags of my training set since i can just grab them when i get the image automatically.

**Comment by asomoza** [comment]
I don't fully understand what you mean with `not have the tokens split into other words, and damage things it already knows`, do you mean `catastrophic forgetting`? That's what LoRAs are for though but common (wrong) practices where people train LoRA's with a rank of 64, 128 or  even 256 makes the model change more of what it already knows too.

Booru tags most of the time are common words so it would be impossible that any training doesn't change what the model already knows, for example the tag `smile` already has 1 token associated to it, so if you train a model with anime images with that tag it would start associating it to the `anime` style more often, that's why people sometimes train dreambooth models with rarely used words like `sks`.

Still what you're probably looking for is [pivotal tuning](https://huggingface.co/blog/sdxl_lora_advanced_script#pivotal-tuning) where you can also train `textual inversion` with the LoRA, you can test it here in diffusers:

https://github.com/huggingface/diffusers/tree/main/examples/advanced_diffusion_training

**Reply by TeKett** [reply]
I want to expand on the number of words/tokens that the model knows. What im looking for is a way to do [this ](https://huggingface.co/docs/transformers/v4.37.2/en/main_classes/tokenizer#transformers.PreTrainedTokenizer.add_tokens) but for a clip model thats already packeged inside of a model. 

From what i understand i can do this with a pipeline object, but i dont have a pipline object. Im already using kohya to do my training, but its kinda pointless when half the words are unknown to the model, so it ends up training some random word thats suposed to be part of an artist name or something similar. Take "slugbox" for example. If i train that, then it will unlearn what it already knows about "slug" and "box". Now do that for 40 thousand tags and you end up damaging around 100 thousand tokens and concepts the model already knows. Even if i use a token like "sks" then thats also bad, since sks is a russian firearm, and imagine the logistics around remembering each and every alias.

But it whould be way more convenient for everyone if we could just add those tokens as new ones, adding new concepts rather then relearning them. Booru tags are multiple words joined into a single word, so its not the same as using the tokens that already exists. "blue hair" and "blue_hair" is not the same thing. 

The former is 2 concepts, wich often cause bleeding, such as making the whole image blue. The latter is a single concept. What you are doing is containing the concept, rather then making the token multi purpose. Making tokens multipurpose has obvious side effects, some of which is the large part in why SD is struggling so much. I understand that this could potentially bloat the model by adding a few tens of thousands of tokens, but i dont really care about that since its only for personal use.

**Reply by asomoza** [reply]
I see what you meant, I don't think its possible but this question should probably be asked in the transformers library, for what you want to do you'll need to first finetune the Clip Model with those new tokens and corresponding images so that the Clip model understand what they are, I never finetuned a Clip Model before so I don't know how much time it takes or how many images you need but I'm guessing a lot.

The next part is what I'm not sure of, if you have a completely different finetuned Clip Model I don't think it would be compatible with the `unet` trained from Stability AI since the embeddings would be different or at the very least you can expect some of them to change which could be worse than just training a LoRA with what you want.

If it somehow it works with a custom finetuned Clip Model, then you can train the SD model with the new tokens. You can't just add new tokens and train a Stable Difussion checkpoint directly.

What I don't understand is: if this is for personal use, why don't you just train multiple LoRA's with the concepts that you want and switch them when you need them?  That's what most of the people do and why there's a huge "market" of LoRAs in sites like CivitAI.

If you really want to train all the tags from danbooru for personal use you'll need to spend a lot of money and time doing it even if you could add new tokens, at that point it would be better to just train a model from scratch with something like Würstchen.

**Reply by TeKett** [reply]
I already asked over on transformers, and they said that it was outside of the scope and that i should ask here.

"The next part is what I'm not sure of, if you have a completely different finetuned Clip Model I don't think it would be compatible with the unet trained from Stability AI" Well thats for step 2, for now im on step 1. How do i even edit an existing checkpint, or make my own checkpoint, with a custom clip model. 

I dont want to train tens of thousands of loras, and i still want the added vocabulary. And its not my goal to make one lora per concept, but ultimetly train a checkpoint that can do everything. Doing one lora per concept still has the logistical problem of cataloging tens of thousands or loras, and storing them, and then using them in a convenient manner. I dont relly want to use like 30 loras at the same time.

"money and time", only time, the money has already been spent, i run locally.

You dont really need to understand, all im asking for is "can" and "how", im not interested in debating the morals and ethics about it.

**Reply by TeKett** [reply]
From what i understand, as i said, i can use a pipeline object. But that is some base pipeline of stable diffusion, but i want to be able to do the same with a distributed custom checkpoint. I guess i could do that if its the only way, unless i could convert the .saftensor into a .ckp, and then convert the .ckp into a diffusers model / pipeline, wich all should be possible i think.

This is the code i got from the people over on transformers.
```
from diffusers import StableDiffusionPipeline, EulerDiscreteScheduler
import torch 

pipeline = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5")

num_new_tokens = pipeline.tokenizer.add_tokens(["new_token_1", "new_token_2"], special_tokens=True)

# simple resize
pipeline.text_encoder.resize_token_embeddings(len(pipeline.tokenizer))

# overwrite the content to have better results
input_embeddings = pipeline.text_encoder.get_input_embeddings().weight.data
input_embeddings_avg = input_embeddings[:-num_new_tokens].mean(dim=0, keepdim=True)
input_embeddings[-num_new_tokens:] = input_embeddings_avg
pipeline.text_encoder.set_input_embeddings(input_embeddings)
```

**Reply by sayakpaul** [reply]
Why do you have to rely on the `pipeline` object for doing this?

**Reply by TeKett** [reply]
Please enlighten me if i dont need to do it that way, since thats the whole point in me even asking. If i knew how to do this then i whould not have asked.

**Reply by asomoza** [reply]
I was just trying to understand why are you so fixated in adding custom tokens to the Clip Model instead of what everyone else does, I really don't care about the morals or ethics behind it.

I could probably help with the "how" but I really don't know if it would work, so I was going to ping @sayakpaul, but he saw this discussion, so I leave it to him, hope it works for you.

**Reply by sayakpaul** [reply]
I don't see any method that is pipeline specific here. Like `get_input_embeddings()` comes from the text encoder model and NOT from the pipeline.

**Reply by TeKett** [reply]
So how do i load up the checkpoint for editing? Remember, im not he one who maintains or writes the project. I do not know what you know, i am not all knowing, you have to tell me what you know in order for me to know.

**Reply by sayakpaul** [reply]
Easy there, buddy. @asomoza is not the maintainer of this repository, either. But they are trying their best to help out. 

To give you an example of how you can load a particular component:

```python
from transformers import CLIPTextModel

text_encoder = CLIPTextModel.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", subfolder="text_encoder")
```

**Reply by TeKett** [reply]
Well you are, hence why i said it. Since as a person with adhd, i know its possible to forget that ppl dont know things you do.

Unintuitive, so, what i want is called from_single_file. I cant see most of the documentaion or most functions with autocomplete in IDE, and half the time huggingface documentation search doesnt work. I cant tell yet if the code does what i want, i got it mostly from the transformers ppl, but it errors out at line 18. `cannot assign 'torch.FloatTensor' as child module 'token_embedding'`

```
from diffusers import StableDiffusionPipeline

array = []
with open("D:/tagstest.txt",encoding="utf8") as file:
        array = [row.rstrip("\n") for row in file.readlines()]

pipeline = StableDiffusionPipeline.from_single_file("C:/Train/checkpoint.safetensors")

num_new_tokens = pipeline.tokenizer.add_tokens(array, special_tokens=False)

# simple resize (is this correct?)
pipeline.text_encoder.resize_token_embeddings(len(pipeline.tokenizer) + num_new_tokens)

# overwrite the content to have better results
input_embeddings = pipeline.text_encoder.get_input_embeddings().weight.data
input_embeddings_avg = input_embeddings[:-num_new_tokens].mean(dim=0, keepdim=True)
input_embeddings[-num_new_tokens:] = input_embeddings_avg
pipeline.text_encoder.set_input_embeddings(input_embeddings) # Error
pipeline.model.save_pretrained("c:/test")
```

**Reply by TeKett** [reply]
And how can i save it as a checkpoint? Since i assume that pipeline.model.save_pretrained() saves it as a diffusers model, and not as a single file.

**Reply by sayakpaul** [reply]
You will have to run your own conversion for getting the checkpoint to a single file-format.

**Reply by TeKett** [reply]
So, i managed to fix the erroring so i think its working, no thanks to the people here. Now i just have to check if its working, and or if i have to train something.

```
from diffusers import StableDiffusionPipeline

array = []
with open("D:/tagstest.txt",encoding="utf8") as file:
        array = [row.rstrip("\n") for row in file.readlines()]

pipeline = StableDiffusionPipeline.from_single_file("C:/Train/ricecakeRemix.safetensors")

num_new_tokens = pipeline.tokenizer.add_tokens(array, special_tokens=False)

# simple resize
pipeline.text_encoder.resize_token_embeddings(len(pipeline.tokenizer))

# overwrite the content to have better results
input_embeddings = pipeline.text_encoder.get_input_embeddings()
input_embeddings_avg = input_embeddings.weight.data[:-num_new_tokens].mean(dim=0, keepdim=True)
input_embeddings.weight.data[-num_new_tokens:] = input_embeddings_avg
pipeline.text_encoder.set_input_embeddings(input_embeddings)
pipeline.save_pretrained("c:/test")
```

**Reply by TeKett** [reply]
Well, when i try to load whatever i got

Diffusers is giving me
```
RuntimeError: Error(s) in loading state_dict for CLIPTextModel:
	size mismatch for text_model.embeddings.token_embedding.weight: copying a param with shape torch.Size([90323, 768]) from checkpoint, the shape in current model is torch.Size([49408, 768]).
```
And A1111 is giving me
```
RuntimeError: Error(s) in loading state_dict for LatentDiffusion:
        size mismatch for cond_stage_model.transformer.text_model.embeddings.token_embedding.weight: copying a param with shape torch.Size([90323, 768]) from checkpoint, the shape in current model is torch.Size([49408, 768]).
```
